### PR TITLE
*: add scope check when get system variables

### PR DESCRIPTION
**Pull request by spongedu**
## What have you changed? (mandatory)

Add scope check when get system variables.
`session` variables cant' be accessed via `SELECT @@global.xxxx`. The same, `global` variables can'e be accessed via `SELECT @@session.xxx` or `SELECT @@local.xxx`.

## What are the type of the changes (mandatory)?

`enhancement`

## How has this PR been tested (mandatory)?

`Unittest`

### CODE REVIEW

File: ast/expressions.go [hunk 1]
```diff
@@ -944,6 +944,8 @@ type VariableExpr struct {
 	IsGlobal bool
 	// IsSystem indicates whether this variable is a system variable in current session.
 	IsSystem bool
+	// ExplicitScope indicates whether this variable scope is explicit set.
```
Comment (shenli) [inline ast/expressions.go]: is set explicitly

File: parser/parser.y [hunk 2]
```diff
@@ -5008,6 +5008,7 @@ SystemVariable:
 	{
 		v := strings.ToLower($1)
 		var isGlobal bool
+	    explicitScope := true
```
Comment (shenli) [inline parser/parser.y]: alignment

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -781,6 +781,13 @@ func (er *expressionRewriter) rewriteVariable(v *ast.VariableExpr) {
 	}
 	var val string
 	var err error
+	if v.ExplicitScope {
```
Comment (shenli) [inline plan/expression_rewriter.go]: Why should we care if it is set explicitly or not?

File: ast/expressions.go (hunk 1 above)
Comment (spongedu) [reply ast/expressions.go]: ok

File: parser/parser.y (hunk 2 above)
Comment (spongedu) [reply parser/parser.y]: ok

File: plan/expression_rewriter.go (hunk 3 above)
Comment (spongedu) [reply plan/expression_rewriter.go]: if the scope constraint is not set explicitly, the behavior is to choose the right scope automatically. So the scope check should't happen here, logically.  For example:
for a `global` variable `max_connections`, `select @@max_connections` works, and for a `session` variable `warning_count`, `select @@warnging_count` works as well.

File: sessionctx/variable/varsutil.go [hunk 4]
```diff
@@ -137,6 +137,25 @@ func SetSessionSystemVar(vars *SessionVars, name string, value types.Datum) erro
...
 }
 
+// ValidateGetSystemVar check if system variable exists and validate it's scope when get system variable.
+func ValidateGetSystemVar(name string, isGlobal bool) error {
+	sysVar, exists := SysVars[name]
+	if !exists {
+		return UnknownSystemVar.GenByArgs(name)
+	}
+	switch sysVar.Scope {
+	case ScopeGlobal, ScopeNone:
```
Comment (zz-jason) [inline sessionctx/variable/varsutil.go]: why handle `ScopeGlobal` and `ScopeNone` together?

Comment (spongedu) [reply sessionctx/variable/varsutil.go]: `ScopeNone` variables are global variables that can't be modified dynamically

File: sessionctx/variable/sysvar.go [hunk 5]
```diff
@@ -69,7 +69,7 @@ const (
 var (
 	UnknownStatusVar   = terror.ClassVariable.New(CodeUnknownStatusVar, "unknown status variable")
 	UnknownSystemVar   = terror.ClassVariable.New(CodeUnknownSystemVar, "unknown system variable '%s'")
-	ErrIncorrectScope  = terror.ClassVariable.New(CodeIncorrectScope, "Incorrect variable scope")
+	ErrIncorrectScope  = terror.ClassVariable.New(CodeIncorrectScope, "Variable '%s' is a %s variable")
```
Comment (jackysp) [inline sessionctx/variable/sysvar.go]: Why not use https://github.com/pingcap/tidb/blob/3db7594fa53c9e2dfb71f186231d3723a817b384/mysql/errname.go#L256 ?

Comment (spongedu) [reply sessionctx/variable/sysvar.go]: Indeed. These error codes can be refined. Will fix

**Comment by spongedu** [comment]
@jackysp PTAL

File: sessionctx/variable/varsutil.go [hunk 6]
```diff
@@ -137,6 +137,25 @@ func SetSessionSystemVar(vars *SessionVars, name string, value types.Datum) erro
 	return vars.SetSystemVar(name, sVal)
 }
 
+// ValidateGetSystemVar check if system variable exists and validate it's scope when get system variable.
```
[the remark this answers is not in the document]
Comment (spongedu) [reply sessionctx/variable/varsutil.go]: ok. will fix

**Pull request merged**